Sources/KituraNet/HTTP/HTTPServerConfigurationHandler.swift

```diff
@@ -0,0 +1,76 @@
+import NIO
+import NIOHTTP1
+import NIOWebSocket
+import LoggerAPI
+import Foundation
+import Dispatch
+import NIOConcurrencyHelpers
+
+internal class HTTPServerConfigurationHandler: ChannelDuplexHandler {
+    // The HTTPServer instance on which this handler is installed
+    var server: HTTPServer
+    let requestSizeLimit: Int?
+    let connectionLimit: Int?
+    var requestSize: Int = 0
+    var connectionCount = 0
+    typealias InboundIn = ByteBuffer
+    typealias OutboundIn = ByteBuffer
+    typealias InboundOut = ByteBuffer
+    typealias OutboundOut = ByteBuffer
+    
+    public init(for server: HTTPServer) {
+        self.server = server
+//        if let requestSizeLimit = server.serverConfig.requestSizeLimit {
+//            self.requestSizeLimit = requestSizeLimit
+//        }
+//        if let connectionLimit = server.serverConfig.connectionLimit {
+//            self.connectionLimit = connectionLimit
+//        }
+        self.requestSizeLimit = server.serverConfig.requestSizeLimit ?? nil
+        self.connectionLimit = server.serverConfig.connectionLimit ?? nil
+    }
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let data = self.unwrapInboundIn(data)
+        requestSize = requestSize + data.readableBytes
+        if let requestSizeLimit1 = requestSizeLimit {
+            if requestSize > requestSizeLimit1 {
+                let statusDescription = HTTP.statusCodes[HTTPStatusCode.requestTooLong.rawValue] ?? ""
+                var discriptionBuffer = ByteBufferAllocator().buffer(capacity: requestSizeLimit!)
+                discriptionBuffer.writeString(statusDescription)
+                context.writeAndFlush(NIOAny(discriptionBuffer),promise: nil)
+                requestSize = 0
+                context.close(mode: .all, promise: nil)
+                return
+            }
+        }
+        context.fireChannelRead(wrapInboundOut(data))
+    }
+    
+    public func channelReadComplete(context: ChannelHandlerContext){
+        requestSize = 0
```

Comment:
this almost certainly does not do what is intended here. `channelReadComplete` is invoked at the end of a 'read burst' not at the end of a message. You want to reset `requestSize` when you receive `.end` in `channelRead`.

Let's imagine the client sends us enough data for about three TCP packets in quick succession. When the first one is received, NIO will be able to read some data and will invoke `channelRead` with the data. NIO however will speculatively try to read more just in case there's more data arriving from the network. In our example here the second of the three packets might have arrived so NIO can indeed read more and invoke `channelRead` again. Same for the third packet. After receiving that, NIO again speculatively tries to read more, now there's nothing more available, NIO will signal the end of the read burst by invoking `channelReadComplete` to signal to you that at this very point in time, we're done with reading.

You can however not reset `requestSize` to `0` in `channelReadComplete`. A client that does something like "send a few kilobytes; wait 100 ms; send a few kilobytes; wait 100 ms; repeat" could then send you arbitrary amounts of data because `channelReadComplete` will be invoked. See also the [docs on `channelReadComplete`](https://apple.github.io/swift-nio/docs/current/NIO/Protocols/_ChannelInboundHandler.html#/s:3NIO22_ChannelInboundHandlerP19channelReadComplete7contextyAA0bD7ContextC_tF)

Reply:
this should be fairly easy to reproduce in a unit test which I would recommend writing.

Reply:
Ideally you test just `HTTPServerConfigurationHandler` in isolation using `EmbeddedChannel` where you are in full control of the events sent. If you need examples, please reach out, @ianpartridge and @pushkarnk  might also be able to help.